zhaquirks/bosch/rbsh_rth0_zb_eu.py

```diff
@@ -0,0 +1,147 @@
+"""Device handler for Bosch RBSH-RTH0-ZB-EU thermostat."""
+
+
+from zigpy.quirks import CustomCluster
+from zigpy.quirks.v2 import CustomDeviceV2, add_to_registry_v2
+import zigpy.types as t
+from zigpy.zcl.clusters.hvac import Thermostat, UserInterface
+from zigpy.zcl.foundation import ZCLAttributeDef
+
+"""Bosch specific thermostat attribute ids."""
+
+# Mode of operation with values BoschOperatingMode.
+OPERATING_MODE_ATTR_ID = 0x4007
+
+# Valve position: 0% - 100%
+VALVE_POSITION_ATTR_ID = 0x4020
+
+# Window open switch (changes to a lower target temperature when on).
+WINDOW_OPEN_ATTR_ID = 0x4042
+
+# Boost preset mode.
+BOOST_ATTR_ID = 0x4043
+
+"""Bosch specific user interface attribute ids."""
+
+# Display on-time (5s - 30s).
+SCREEN_TIMEOUT_ATTR_ID = 0x403A
+
+# Display brightness (0 - 10).
+SCREEN_BRIGHTNESS_ATTR_ID = 0x403B
+
+
+class BoschOperatingMode(t.enum8):
+    """Bosh operating mode attribute values."""
+
+    Schedule = 0x00
+    Manual = 0x01
+    Pause = 0x05
+
+
+class State(t.enum8):
+    """Binary attribute (window open) value."""
+
+    Off = 0x00
+    On = 0x01
+
+
+class BoschThermostatCluster(CustomCluster, Thermostat):
+    """Bosch thermostat cluster."""
+
+    class AttributeDefs(Thermostat.AttributeDefs):
+        """Bosch thermostat manufacturer specific attributes."""
+
+        operating_mode = ZCLAttributeDef(
+            id=t.uint16_t(OPERATING_MODE_ATTR_ID),
+            type=BoschOperatingMode,
+            is_manufacturer_specific=True,
+        )
+
+        pi_heating_demand = ZCLAttributeDef(
+            id=t.uint16_t(VALVE_POSITION_ATTR_ID),
+            # Values range from 0-100
+            type=t.enum8,
+            is_manufacturer_specific=True,
+        )
+
+        window_open = ZCLAttributeDef(
+            id=t.uint16_t(WINDOW_OPEN_ATTR_ID),
+            type=State,
+            is_manufacturer_specific=True,
+        )
+
+        boost = ZCLAttributeDef(
+            id=t.uint16_t(BOOST_ATTR_ID),
+            type=State,
+            is_manufacturer_specific=True,
+        )
+
+
+class BoschUserInterfaceCluster(CustomCluster, UserInterface):
+    """Bosch UserInterface cluster."""
+
+    class AttributeDefs(UserInterface.AttributeDefs):
+        """Bosch user interface manufacturer specific attributes."""
+
+        display_on_time = ZCLAttributeDef(
+            id=t.uint16_t(SCREEN_TIMEOUT_ATTR_ID),
+            # Usable values range from 5-30
+            type=t.enum8,
+            is_manufacturer_specific=True,
+        )
+
+        display_brightness = ZCLAttributeDef(
+            id=t.uint16_t(SCREEN_BRIGHTNESS_ATTR_ID),
+            # Values range from 0-10
+            type=t.enum8,
+            is_manufacturer_specific=True,
+        )
+
+
+class BoschThermostat(CustomDeviceV2):
+    """Bosch thermostat custom device."""
```

Comment:
This can be removed. zigpy's `add_to_registry_v2` creates a `CustomDeviceV2` internally.

Reply:
Done